smach_ros/src/smach_ros/introspection.py

```diff
@@ -60,7 +60,7 @@ def set_initial_state(self,
         initial_status_msg = SmachContainerInitialStatusCmd(
                 path = path,
                 initial_states = initial_states,
-                local_data = pickle.dumps(initial_userdata._data,2))
+                local_data = pickle.dumps(initial_userdata._data,2).decode('latin1'))
```

Comment:
I'm surprised to see `encode()` and `decode()`. What was the error before? Why `latin1` instead of the default encoding?

Reply:
I don't really know if this is necessary, maybe it's my pc configuration. Can you try it ?

Reply:
Fixed in a6835a7

It looks like in Python 2.7 `pickle.dumps()` returns a `str`, but in Python 3 it returns a `bytes`. The output is being passed to a field in a ROS message that expects a string. When publishing rospy assumes the field is a `str` and tries to encode it as a bytes object usint the `utf-8` encoding, which obviously won't work since the field is already set to bytes.

```
[smach_ros.rosunit-introspection/test_introspection][ERROR]---------------------
'bytes' object has no attribute 'encode'
  File "/usr/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/usr/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/home/sloretz/ws/noetic/src/executive_smach/smach_ros/test/introspection_test.py", line 79, in test_introspection
    init_set = intro_client.set_initial_state('intro_test','/intro_test',['SM2'],injected_ud,timeout = rospy.Duration(10.0))
  File "/home/sloretz/ws/noetic/src/executive_smach/smach_ros/src/smach_ros/introspection.py", line 91, in set_initial_state
    init_pub.publish(initial_status_msg)
  File "/opt/ros/noetic/lib/python3/dist-packages/rospy/topics.py", line 882, in publish
    self.impl.publish(data)
  File "/opt/ros/noetic/lib/python3/dist-packages/rospy/topics.py", line 1066, in publish
    serialize_message(b, self.seq, message)
  File "/opt/ros/noetic/lib/python3/dist-packages/rospy/msg.py", line 152, in serialize_message
    msg.serialize(b)
  File "/home/sloretz/ws/noetic/devel_isolated/smach_msgs/lib/python3/dist-packages/smach_msgs/msg/_SmachContainerInitialStatusCmd.py", line 84, in serialize
    _x = _x.encode('utf-8')
--------------------------------------------------------------------------------
```


Encoding/decoding the bytes object doesn't seem to be the right thing to do here. I tried changing the encoding to `utf-8` so the encode call when serializing the message wouldn't change anything, but the pickle'd object can't be trusted to be a valid utf-8 string. I see an exception `'utf-8' codec can't decode byte 0x80 in position 0: invalid start byte` with this strategy. 

Instead in a6835a7 when creating the message it base64 encodes the pickle data to get a `bytes` object that is definitely valid `utf-8` and then encodes that. When going the other direction it only needs to `b64decode()` before unpickling since `b64decode()` accepts strings.